### ISSUE TITLE
Python fails to read a script whose path is `/dev/fd/X`

### ISSUE THREAD

**Issue by b50f524d-56e2-4a9a-a7b2-5a117a960b6c** [python/cpython#87235]
BPO | [43069](https://bugs.python.org/issue43069)
--- | :---
Nosy | @ronaldoussoren, @ned-deily, @izbyshev, @cipriancraciun

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-01-29.22:32:21.638>
labels = ['OS-mac', 'interpreter-core', '3.8', '3.9']
title = 'Python fails to read a script whose path is `/dev/fd/X`'
updated_at = <Date 2021-01-30.12:21:41.381>
user = 'https://github.com/cipriancraciun'
```

bugs.python.org fields:
```python
activity = <Date 2021-01-30.12:21:41.381>
actor = 'izbyshev'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Interpreter Core', 'macOS']
creation = <Date 2021-01-29.22:32:21.638>
creator = 'ciprian.craciun'
dependencies = []
files = []
hgrepos = []
issue_num = 43069
keywords = []
message_count = 3.0
messages = ['385955', '385978', '385982']
nosy_count = 4.0
nosy_names = ['ronaldoussoren', 'ned.deily', 'izbyshev', 'ciprian.craciun']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue43069'
versions = ['Python 3.8', 'Python 3.9']
```

</p></details>

### Linked PRs
* gh-99768
* gh-99816
* gh-99817

**Comment by ronaldoussoren** [comment]
The problem is not in pythonrun.c as the program doesn't get there. The issue is very weird, and could be a bug in the OS.

The script is read by Modules/main.c: pymain_run_file_obj. That calls ``_Py_fopen_obj`` to open the file object, and when I instrument pymain_run_file_obj I see that the file offset is wrong: when the contents of ``/dev/fd/9`` is large enough the result of ``ftell(3)`` for the newly opened file is not 0 but at EOF. Calling ``rewind(3)`` just after opening "fixes" the issue, but that's a gross hack and not the correct fix.

```diff

git diff ../Modules/main.c                                                                                                                         (gh-83765-doc-large-shmmem)cpython
diff --git a/Modules/main.c b/Modules/main.c
index 7edfeb3365..d5a7577098 100644
--- a/Modules/main.c
+++ b/Modules/main.c
@@ -329,6 +329,8 @@ pymain_run_file_obj(PyObject *program_name, PyObject *filename,
                            program_name, filename, errno, strerror(errno));
         return 2;
     }
+    PySys_FormatStderr("opened %R at %ld\n", filename, ftell(fp));
+    rewind(fp);
 
     if (skip_source_first_line) {
         int ch;
```

With this patch a small file prints: ```opened '/dev/fd/9' at 0```, while a larger file prints ``opened '/dev/fd/9' at 72`` (file size is 72 bytes). 

Tracing further: _Py_fopen_obj is defined in Python/fileutils.c, and calls fopen (on unix-y systems like macOS). Similar instrumentation in that file shows that the file offset is wrong directly after calling fopen(3). 

But that's in the context of Python, a small program that contains similar code works fine.  However...  If I change that program to first read some binds and then reopen /dev/fd/... the second open inherits the file offset from the first file pointer!

E.g. (in pseudo C without error checking):

```
FILE* fp1 = fopen("/dev/fd/9", "rb");
fgets(fp1);
FILE* fp2 = fopen("/dev/fd/9", "rb");
tell(fp2); // <--- Does *not* return 0
```

I do not yet know how relevant this is, from what I've seen so far ``pymain_run_file_obj`` is the first time the file gets opened.

**Comment by ronaldoussoren** [comment]
I've done some more sleuthing, and the finding in the previous message was relevant...

Turns out the file is opened a second time, but not using ``_Py_fopen_obj``: pymain_run_python calls pymain_get_importer, which calls PyImport_GetImporter and that calls into importlib, which call zip import.zipimport which opens the file and tries to read a zip directory from it.  This changes the file offset and does not reset it (and that's generally not necessary).

This also explains why small files work fine: those are to small to seek to the offset of the zip directory, and hence zipimporter fails when it tries to change the file offset.

In a quick hack I changed zipimporter._read_directory to reset the file offset and that fixes this issue:

```diff

diff --git a/Lib/zipimport.py b/Lib/zipimport.py
index 016f1b8a79..9d0e6c808d 100644
--- a/Lib/zipimport.py
+++ b/Lib/zipimport.py
@@ -347,6 +347,8 @@ def _read_directory(archive):
         raise ZipImportError(f"can't open Zip file: {archive!r}", path=archive)
 
     with fp:
+      ro_off = fp.tell()
+      try:
         try:
             fp.seek(-END_CENTRAL_DIR_SIZE, 2)
             header_position = fp.tell()
@@ -455,6 +457,8 @@ def _read_directory(archive):
             t = (path, compress, data_size, file_size, file_offset, time, date, crc)
             files[name] = t
             count += 1
+      finally:
+          fp.seek(ro_off, 0)
     _bootstrap._verbose_message('zipimport: found {} names in {!r}', count, archive)
     return files
 
```

Inline patch because it is not ready for a PR at this point.

**Comment by ronaldoussoren** [comment]
Summary:

- On macOS the file offset for multiple open file descriptors for ``/dev/fd/N`` are not independent.
- ``python /dev/fd/N`` opens ``/dev/fd/N`` twice: first in zipimport and then to actually read the script. Zipimport changes the file offset and the code that actually runs the script sees an empty file (for a large enough file)
- A workaround for this issue is to change zipimporter to reset the file offset when scanning the zip directory

Creating a PR for this shouldn't be too hard, but not today...   This definitely needs a test.

**Comment by ronaldoussoren** [comment]
I've added a PR that fixes the issue, including a test that failed before I changed the zipimport module.

**Comment by vstinner** [comment]
The newly added test fails on FreeBSD, see issue #100005. I proposed PR #100006 to fix the issue.

**Comment by hauntsaninja** [comment]
Nice catch! Looks like the 3.10 backport is still pending

**Comment by ambv** [comment]
Fixed in 3.10 as well.
